src/libstore/derivation-options.hh

```diff
@@ -0,0 +1,171 @@
+#pragma once
+///@file
+
+#include <cstdint>
+#include <nlohmann/json.hpp>
+#include <optional>
+#include <variant>
+
+#include "types.hh"
+#include "json-impls.hh"
+
+namespace nix {
+
+class Store;
+struct BasicDerivation;
+class ParsedDerivation;
+
+/**
+ * This represents all the special options on a `Derivation`.
+ *
+ * Currently, these options are parsed from the environment variables
+ * with the aid of `ParsedDerivation`.
+ *
+ * The first goal of this data type is to make sure that no other code
+ * uses `ParsedDerivation` to ad-hoc parse some additional options. That
+ * ensures this data type is up to date and fully correct.
+ *
+ * The second goal of this data type is to allow an alternative to
+ * hackily parsing the options from the environment variables. The ATerm
+ * format cannot change, but in alternatives to it (like the JSON
+ * format), we have the option of instead storing the options
+ * separately. That would be nice to separate concerns, and not make any
+ * environment variable names magical.
+ */
```

Comment:
Is it meant to be a lossless representation of part of the derivation, or is it supposed to represent part of the derivation semantically uniquely?

Reply:
Ah yes it is supposed to reflect the semantics, not the syntax.

In other words, there will not be a `DerivationOptions  -> Env` function; instead, the ATerm logic will (eventually):

1. Check that the options are implied by the rest of the `Derivation`
2. Ignore the (eventual new) `DerivationOptions options;` field of the derivation
3. Serialize as today

So yes let's do these changes.